// SheetClasses/AOCDivisionSpreadsheet.js
var GWLSpreadsheet = require('@sheets/GWLSpreadsheet');
var GWLDivisionSpreadsheet = require('@sheets/GWLDivisionSpreadsheet')
var googleAuth = require('@mymodules/auth');
const localfs = require('@mymodules/localfs');
const utils = require('@mymodules/utils');
const sanitize = require('@mymodules/utils').sanitize;
var {google} = require('googleapis');
var fs = require('fs');
var sheetsApi = google.sheets('v4');

module.exports = class AOCDivisionSpreadsheet extends GWLDivisionSpreadsheet {
    // initializes an instance of this class, override the format, class name, 
	// teampage path, and anything else
	constructor(name, internal, ongoing, spreadsheetId, numMaps) {
		super(name, internal, ongoing, spreadsheetId, numMaps);
		
		// Change these!
		this.format = "ow2_division";
		this.className = "AOCDivisionSpreadsheet";
		this.teamPageTemplatePath = "/client/Teams/team_division.html";
		this.meta.hasStats = true;
		this.meta.hasPlayerProfiles = true;
		this.meta.hasTournaments = true;
		this.tournaments = [];
	}

    // loads a serialized object into the class
	static loadObject(obj) {
		return new Promise((resolve,reject) => {
			if(obj.format != "ow2_division") {
				console.log("Object is not a GWLDivisionSpreadsheet object!");
				reject("InvalidFormat");
			}
			var sheet = new AOCDivisionSpreadsheet(obj.name, obj.internal, obj.ongoing, obj.spreadsheetId);
			sheet.teams = obj.teams;
			sheet.tournaments = obj.tournaments;
			console.log(sheet.name, sheet.internal, sheet.spreadsheetId);
			resolve(sheet);
		});
	}
}
