// SheetClasses/GWLDivisionSpreadsheet.js
var GWLSpreadsheet = require('@sheets/GWLSpreadsheet');
var googleAuth = require('@mymodules/auth');
const localfs = require('@mymodules/localfs');
const utils = require('@mymodules/utils');
const sanitize = require('@mymodules/utils').sanitize;
var {google} = require('googleapis');
var fs = require('fs');
var sheetsApi = google.sheets('v4');

// =======================
// How to add a sheet:
//
// 1. Duplicate this class, go through and add code where comments tell unless
// 2. Go to spreadsheet-types.js and add a require() for it, then add code to
//    loadSheet() that checks for the format and adds it. 
// =======================

// TODO name class
module.exports = class GWLDivisionSpreadsheet extends GWLSpreadsheet {
	// initializes an instance of this class, override the format, class name, 
	// teampage path, and anything else
	constructor(name, internal, ongoing, spreadsheetId) {
		super(name, internal, ongoing, spreadsheetId);
		
		// Change these!
		this.format = "division";
		this.className = "GWLDivisionSpreadsheet";
		this.teamPageTemplatePath = "/client/Teams/team_division.html";
		this.meta.hasStats = true;
		this.meta.hasPlayerProfiles = true;
		this.meta.hasTournaments = true;
		this.tournaments = [];
	}
	
	// loads a serialized object into the class
	static loadObject(obj) {
		return new Promise((resolve,reject) => {
			if(obj.format != "division") {
				console.log("Object is not a GWLDivisionSpreadsheet object!");
				reject("InvalidFormat");
			}
			var sheet = new GWLDivisionSpreadsheet(obj.name, obj.internal, obj.ongoing, obj.spreadsheetId);
			sheet.teams = obj.teams;
			sheet.tournaments = obj.tournaments;
			console.log(sheet.name, sheet.internal, sheet.spreadsheetId);
			resolve(sheet);
		});
	}
	
	// builds the directory structure in ./data/[internal]
	// returns promise that resolves when local directory is complete and
	// rejects if there is an fs error.
	// by default creates a ./data/[internal] and ./data/[internal]/teams directory, but can be overridden
	buildDirectory() {
		return new Promise((resolve,reject) => {
			fs.mkdirSync("./data/" + this.internal);
			fs.mkdirSync("./data/" + this.internal + "/teams");
			fs.mkdirSync("./data/" + this.internal + "/tournaments");
			
			// don't cache if this is an ongoing tournament.
			if(this.ongoing) {
				fs.copyFileSync('./gitignore_template', './data/' + this.internal + '/teams/.gitignore');
				fs.copyFileSync('./gitignore_template', './data/' + this.internal + '/.gitignore');
				fs.copyFileSync('./gitignore_template', './data/' + this.internal + '/tournaments/.gitignore');
			}
			resolve(null);
		});
	}
	
	// called by build function to get season info from the sheet's Info tab
	// this loads metadata required to handle the season's info, such as team names
	loadSheetInfo() {
		return new Promise((resolve,reject) => {
			googleAuth.authorize()
			.then((auth) => {
				sheetsApi.spreadsheets.values.batchGet({
					auth: auth,
					spreadsheetId: this.spreadsheetId,
					ranges: [
						'Info!A2:F21',
						'Info!I6:I16',
						
					]
				}, (err, response) => {
					if(err) {
						console.log("Google API returned an error: " + err);
						reject(err);
					}
					
					// store teams
					let teams = response.data.valueRanges[0].values;
					if(teams) teams.map((row) => {
						let obj = {};
						obj.name = row[0];
						obj.division = row[1];
						obj.internal = row[2];
						obj.iconUrl = row[3];
						obj.primaryColor = row[4];
						obj.secondaryColor = row[5];
						
						// discard if team is invalid (no name or internal name)
						if(obj.name != undefined && obj.name != '' && obj.internal != undefined && obj.internal != '') {
							console.log(`pushing team ${obj.name} to ${this.name}`);
							this.teams.push(obj);
						}
					});
					
					// store tournaments
					let tournaments = response.data.valueRanges[1].values;
					if(tournaments) tournaments.map((row) => {
						this.tournaments.push({ name: row[0], internal: row[0].split(' ').join('_') }); // append tournament
					});
					console.log(this.tournaments);
					
					// batch get completed and results stored, we can resolve now
					resolve(this);
				});
			})
			.catch((err) => {
				// called when authorize() fails
				console.error('Authentication error: ', err);
				reject(err);
			});
		});
	}
	
	// stores a computed link dictionary
	storeLinkDict(dict) {
		return new Promise((resolve,reject) => {
			// store the dictionary
			// the path can be changed but you should also update the getLinkDict function below
			localfs.writeJsonPromise(`./data/${this.internal}/linkdict.json`, dict)
			.then(() => resolve("Success!"))
			.catch(err => {
				console.log("Error storing dictionary for sheet ", this.internal);
				reject(err);
			});
		});
	}
	
	// retrieves the link dictionary, see above
	getLinkDict() {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise(`./data/${this.internal}/linkdict.json`)
			.then(obj => resolve(obj))
			.catch(err => reject(err)); // should not happen unless someone manually entered an invalid dictionary
		});
	}
	
	// runs a BatchGet Google API request and stores the results in whatever
	// local file format the sheet uses. This is where most of the work is done. 
	batchGetAll() {
		return new Promise((resolve,reject) => {
			// calculate length of standings
			let lastStandingsRow = this.teams.length + 22;
			// build ranges
			let ranges = [];
			ranges.push('MatchLog!A2:AB361'); // match log
			ranges.push('HeroStats!A2:J2000'); // hero stats
			ranges.push('HeroStats!L2:T33'); // avg hero stats
			ranges.push('HeroStats!N34:S35'); // league averages
			ranges.push(`Standings!A23:O${lastStandingsRow}`); // standings
			ranges.push('PlayerInfo!A2:AC200'); // player info
			ranges.push('Info!H3'); // current week
			for(let i=0; i<this.teams.length; i++) {
				ranges.push(this.teams[i].internal + '!A5:L14') // roster
				ranges.push(this.teams[i].internal + '!F70:J88') // map stats
				ranges.push(this.teams[i].internal + '!A32:Y67') // match history
				ranges.push(this.teams[i].internal + '!J17:U20') // team stats
			}
			for(let i=0;i<this.tournaments.length; i++) {
				ranges.push('\'' + this.tournaments[i].name + '\'!A2:AD34') // push tournaments
			}
			console.log(ranges);
			
			// push request for these ranges
			googleAuth.authorize()
			.then((auth) => {
				sheetsApi.spreadsheets.values.batchGet({
					auth: auth,
					spreadsheetId: this.spreadsheetId,
					ranges: ranges
				}, (err, response) => {
					// handle error
					if(err) {
						console.error('sheetsApi.batchGet returned an error: ', err);
						console.error(err.stack);
						reject(err);
					}
					
					// handle the data returned
					// not strictly required but I recommend handling in a separate storeBatchGet function
					this.storeBatchGet(response.data)
					// don't forget to resolve!
					.then(res => resolve(res))
					.catch(err => reject(err));
				});
			})
			.catch((err) => {
				// called when authorize() fails
				console.error('Authentication error: ', err);
				reject(err);
			});
		});
	}
	
	// stores data captured by batchGetAll
	// again, not strictly necessary but helps avoid callback hell
	storeBatchGet(data) {
		return new Promise((resolve,reject) => {
			// highly recommend to make separate private functions for these as well,
			// using either a promise chain or a Promise.all()
			
			// TODO store the current round
			this.currentRound = parseInt(data.valueRanges[6].values[0][0]);
			// TODO store the match log
			this._storematchlog(data.valueRanges[0].values)
			// TODO store hero stats
			.then(() => this._storeherostats(data.valueRanges[1].values, data.valueRanges[2].values, data.valueRanges[3].values))
			// TODO store standings
			.then(() => this._storestandings(data.valueRanges[4].values))
			// TODO store player info
			.then(() => this._storeplayerinfo(data.valueRanges[5].values))
			// TODO store team pages
			.then(() => this._storeteaminfo(data))
			.then(() => this._storetournamentinfo(data))
			.then(res => {
				console.log('Batch get for ' + this.name + ' complete!');
				resolve(this);
			})
			.catch(err => {
				console.log("Error storing BatchGet data! \n" + err);
				reject(err);
			});
		});
	}
	
	// ====== Put private functions used in data storage below! ======== 
	_storematchlog(range) {
		return new Promise((resolve,reject) => {
			// store matchlog info
			{
				var json = '{"currentround":';
				json += this.currentRound;
				json += ',"matches":[';
				if(range && range.length) {
					range.map((row) => {
						json += `{"tournament":"${sanitize(row[0])}","played":"${sanitize(row[7])}",`;
						json += `"team1":"${sanitize(row[1])}","team2":"${sanitize(row[2])}",`;
						json += `"division":"${sanitize(row[4])}","date":"${sanitize(row[5])}","time":"${sanitize(row[6])}",`;
						json += `"map1":{"name":"${sanitize(row[8])}","winner":"${sanitize(row[9])}"},`;
						json += `"map2":{"name":"${sanitize(row[10])}","winner":"${sanitize(row[11])}"},`;
						json += `"map3":{"name":"${sanitize(row[12])}","winner":"${sanitize(row[13])}"},`;
						json += `"map4":{"name":"${sanitize(row[14])}","winner":"${sanitize(row[15])}"},`;
						json += `"map5":{"name":"${sanitize(row[16])}","winner":"${sanitize(row[17])}"},`;
						json += `"map6":{"name":"${sanitize(row[18])}","winner":"${sanitize(row[19])}"},`;
						json += `"map7":{"name":"${sanitize(row[20])}","winner":"${sanitize(row[21])}"},`;
						json += `"map8":{"name":"${sanitize(row[22])}","winner":"${sanitize(row[23])}"},`;
						json += `"matchwinner":"${sanitize(row[24])}","vod":"${sanitize(row[25])}","round":"${sanitize(row[26])}"},`;
					});
					
					json = json.replace(/,$/,'');
					json += ']}';
				} else {
					console.log("Pulled an empty MatchLog!");
					json += ']}';
				}
					
				let jsonObj = JSON.parse(json);
				localfs.writeJsonPromise('./data/' + this.internal + '/matchlog.json', jsonObj)
				.then(() => resolve(jsonObj))
				.catch(err => {
					console.log("Error storing matchlog!");
					reject(err);
				});
			}
		});
	}
	
	_storeherostats(rows,league,totals) {
		return new Promise((resolve,reject) => {
			var json = '{"stats":[';
			if(rows && rows.length && league && league.length && totals && totals.length) {
				rows.map((row) => {
					json += `{"player":"${sanitize(row[0])}","hero":"${sanitize(row[1])}",`;
					json += `"elims":"${row[2]}","fb":"${row[3]}",`;
					json += `"damage":"${row[4]}","deaths":"${row[5]}",`;
					json += `"healing":"${row[6]}","blocked":"${row[7]}",`;
					json += `"timeplayed":"${row[8]}","team":"${row[9]}"},`;
				});
				
				league.map((row) => {
					json += `{"player":"${sanitize(row[0])}","hero":"${sanitize(row[1])}",`;
					json += `"elims":"${row[2]}","fb":"${row[3]}",`;
					json += `"damage":"${row[4]}","deaths":"${row[5]}",`;
					json += `"healing":"${row[6]}","blocked":"${row[7]}",`;
					json += `"timeplayed":"${row[8]}","team":"The Gopherwatch League"},`;
				});
				
				json = json.replace(/,$/,'');
				json += '],"averages":{';
				json += `"elims":"${parseFloat(totals[0][0]).toFixed(2)}","elims10":"${parseFloat(totals[1][0]).toFixed(2)}","fb":"${parseFloat(totals[0][1]).toFixed(2)}","fb10":"${parseFloat(totals[1][1]).toFixed(2)}",`;
				json += `"dmg":"${parseFloat(totals[0][2]).toFixed(2)}","dmg10":"${parseFloat(totals[1][2]).toFixed(2)}","deaths":"${parseFloat(totals[0][3]).toFixed(2)}","deaths10":"${parseFloat(totals[1][3]).toFixed(2)}",`;
				json += `"healing":"${parseFloat(totals[0][4]).toFixed(2)}","healing10":"${parseFloat(totals[1][4]).toFixed(2)}","blocked":"${parseFloat(totals[0][5]).toFixed(2)}","blocked10":"${parseFloat(totals[1][5]).toFixed(2)}"}}`;
			} else {
				console.log("Pulled an empty HeroStats!");
				json += ']}';
			}
			
			let jsonObj = JSON.parse(json);
			localfs.writeJsonPromise('./data/' + this.internal + '/herostats.json', jsonObj)
			.then(() => resolve(jsonObj))
			.catch(err => {
				console.log("Error storing HeroStats!");
				reject(err)
			});
		});
	}
	
	_storestandings(range) {
		return new Promise((resolve,reject) => {
			let json = '{"Teams":[';
			if(range && range.length) {
				range.map((row) => {
					json += `{"name":"${sanitize(row[1])}","rank":"${sanitize(row[0])}","division":"${sanitize(row[2])}","points":"${sanitize(row[3])}",`;
					json += `"win":"${sanitize(row[4])}","loss":"${sanitize(row[5])}",`;
					if(row[3]+row[4]==0) {
						json += `"pct":"0%",`;
					} else {
						let winrate = parseInt(row[4])+parseInt(row[5]);
						winrate = parseInt(row[4])/winrate*100;
						winrate = winrate.toFixed(2);
						json += `"pct":"${winrate}%",`;
					}
					json += `"mapwin":"${sanitize(row[7])}","maploss":"${sanitize(row[8])}","maptie":"${sanitize(row[9])}",`;
					json += `"mapdiff":"${sanitize(row[10])}","dwin":"${sanitize(row[11])}","dloss":"${sanitize(row[12])}",`;
					json += `"ndwin":"${sanitize(row[13])}","ndloss":"${sanitize(row[14])}"},`;
				});
				
				json = json.replace(/,$/,'');
				json += ']}';
			} else {
				console.log("Pulled an empty standings!");
				json += ']}';
			}

			var obj = JSON.parse(json);
			localfs.writeJsonPromise('./data/' + this.internal + '/standings.json', obj)
			.then(() => resolve(obj))
			.catch(err => {
				console.log("Error storing standings!");
				reject(err);
			});
		});
	}
	
	_storeplayerinfo(range) {
		return new Promise((resolve,reject) => {
			let json = '{"season":"';
			json += this.name;
			json += '","players":['
			if(range && range.length) {
				range.map((row) => {
					json += '{"battletag":"';
					json += `${sanitize(row[0])}`;
					json += '","team":"';
					json += `${sanitize(row[1])}`;
					json += '","draft":"';
					json += `${sanitize(row[2])}`;
					json += '","tank":"';
					json += `${sanitize(row[9])}`;
					json += '","dps":"';
					json += `${sanitize(row[10])}`;
					json += '","support":"';
					json += `${sanitize(row[11])}`;
					json += `","mvp":"${sanitize(row[12])}","realname":"${sanitize(row[13])}","playernumber":"${sanitize(row[14])}",`;
					json += `"pronouns":"${sanitize(row[15])}","hometown":"${sanitize(row[16])}","major":"${sanitize(row[17])}",`;
					json += `"twitch":"${sanitize(row[18])}","twitter":"${sanitize(row[19])}","youtube":"${sanitize(row[20])}","instagram":"${sanitize(row[21])}","reddit":"${sanitize(row[22])}",`;
					let accolades = row[25];
					if(accolades != undefined) {
						accolades = accolades.split("\n").join("\\n");
					}
					let bio = row[27];
					if(bio != undefined) {
						bio = bio.split("\"").join("\\\"");
						bio = bio.split("\n").join("\\n");
					}
					json += `"role":"${sanitize(row[23])}","hero":"${sanitize(row[24])}","accolades":"${sanitize(row[25])}","picture":"${sanitize(row[26])}","bio":"${sanitize(row[27])}"},`;
				});
				
				json = json.replace(/,$/,'');
				json += ']}';
				
				json = json.split("\"undefined\"").join("\"\"");
			} else {
				console.log('Player info is empty!');
				json += ']}';
			}
			
			let obj = JSON.parse(json);
			localfs.writeJsonPromise('./data/' + this.internal + '/players.json', obj)
			.then(() => {
				console.log("Wrote to './data/' + this.internal + '/players.json'!");
				resolve(obj);
			})
			.catch(err => {
				console.log("Error storing player info!");
				reject(err);
			});
		});
	}
	
	_storeteaminfo(data) {
		return new Promise((resolve,reject) => {
			let offset = 7; // what index team info starts
			for(let i=0; i<this.teams.length; i++) {
				// check the value ranges for the team exist
				if(!data.valueRanges[offset+4*i+3]) {
					reject('Invalid batchGetAll response! data.valueRanges[' + (offset+4*i+3) + '] does not exist!');
				}
				
				// map to variables
				let filename = './data/' + this.internal + '/teams/' + this.teams[i].internal + '.json';
				let roster = data.valueRanges[offset+4*i].values;
				let maps = data.valueRanges[offset+4*i+1].values;
				let matches = data.valueRanges[offset+4*i+2].values;
				let stats = data.valueRanges[offset+4*i+3].values;
				
				// store team name
				let json = '{"name":"';
				json += this.teams[i].name;
				json += '","division":"';
				json += this.teams[i].division;
				json += '","color1":"';
				json += this.teams[i].primaryColor;
				json += '","color2":"';
				json += this.teams[i].secondaryColor;
				json += '","roster":[';
				
				// store roster
				if(roster && roster.length) {
					roster.map((row) => {
						json += '{"name":"';
						json += `${sanitize(row[0])}`;
						json += '","draft":"';
						json += `${sanitize(row[1])}`;
						json += '","tank":"';
						json += `${sanitize(row[2])}`;
						json += '","damage":"';
						json += `${sanitize(row[3])}`;
						json += '","support":"';
						json += `${sanitize(row[4])}`;
						json += '"},';
					});
					
					json = json.replace(/,$/,'');
					json += '],';
				} else {
					json += '],';
				}
				
				// store maps
				json += '"maps":[';
				if(maps && maps.length) {
					maps.map((row) => {
						json += '{"mapname":"';
						json += `${sanitize(row[0])}`;
						json += '","wins":"';
						json += `${sanitize(row[1])}`;
						json += '","losses":"';
						json += `${sanitize(row[2])}`;
						json += '","draws":"';
						json += `${sanitize(row[3])}`;
						json += '","winrate":"';
						json += `${sanitize(row[4])}`;
						json += '"},';
					});
						
					json = json.replace(/,$/,'');
					json += '],';
				} else {
					json += '],';
				}
				
				// store matches 
				json += '"matches":[';
				if(matches && matches.length) {
					matches.map((row) => {
						json += '{"tournament":"';
						json += `${sanitize(row[0])}`;
						json += '","opponent":"';
						json += `${sanitize(row[1])}`;
						json += '","date":"';
						json += `${sanitize(row[2])}`;
						json += '","time":"';
						json += `${sanitize(row[3])}`;
						json += '","played":"';
						json += `${sanitize(row[4])}",`;
						json += `"map1":{"name":"${sanitize(row[5])}","winner":"${sanitize(row[6])}"},`;
						json += `"map2":{"name":"${sanitize(row[7])}","winner":"${sanitize(row[8])}"},`;
						json += `"map3":{"name":"${sanitize(row[9])}","winner":"${sanitize(row[10])}"},`;
						json += `"map4":{"name":"${sanitize(row[11])}","winner":"${sanitize(row[12])}"},`;
						json += `"map5":{"name":"${sanitize(row[13])}","winner":"${sanitize(row[14])}"},`;
						json += `"map6":{"name":"${sanitize(row[15])}","winner":"${sanitize(row[16])}"},`;
						json += `"map7":{"name":"${sanitize(row[17])}","winner":"${sanitize(row[18])}"},`;
						json += `"map8":{"name":"${sanitize(row[19])}","winner":"${sanitize(row[20])}"},`;
						json += `"winner":"${sanitize(row[21])}",`;
						json += `"division":"${sanitize(row[22])}",`;
						json += `"vod":"${sanitize(row[23])}",`;
						json += `"round":"${sanitize(row[24])}"},`;
					});
						
					json = json.replace(/,$/,'');
					json += '],';
				} else {
					json += '],';
				}
				
				// store stats
				json += '"stats":{';
				if(stats && stats.length) {
					stats.map((row) => {
						json += `"${row[0]}":{`;
						json += `"wins":"${row[1]}","losses":"${row[2]}","mapwins":"${row[3]}","maplosses":"${row[4]}","mapties":"${row[5]}","mapdiff":"${row[6]}","rank":"${row[11]}"},`;
					});
					
					json = json.replace(/,$/,'');
					json += '}}';
				} else {
					json += '}}';
				}
				
				let jsonObj = JSON.parse(json);
				localfs.writeJsonPromise(filename, jsonObj)
				.then(() => {})
				.catch(err => {
					console.log("Error storing team info!");
					reject(err);
				});
			}
			
			resolve("Done!");
		});
	}
	
	// stores tournament info
	_storetournamentinfo(data) {
		return new Promise((resolve,reject) => {
			let offset = 7 + this.teams.length * 4; // offset tournaments start at
			for(let i=0; i<this.tournaments.length; ++i) {
				// map to variables
				let filename = './data/' + this.internal + '/teams/' + this.tournaments[i].internal + '.json';
				let matches = data.valueRanges[offset+i].values;
				
				// start json
				let json = `{"tournament":"${this.tournaments[i].name}","matches":[`;
				
				// store matches
				if(matches && matches.length) {
					matches.map((row) => {
						json += '{"matchid":"';
						json += `${sanitize(row[0])}`;
						json += '","bracketname":"';
						json += `${sanitize(row[1])}`;
						json += '","parent":"';
						json += `${sanitize(row[2])}`;
						json += '","team1":"';
						json += `${sanitize(row[4])}`;
						json += '","team2":"';
						json += `${sanitize(row[5])}`;
						json += '","date":"';
						json += `${sanitize(row[8])}`;
						json += '","time":"';
						json += `${sanitize(row[9])}`;
						json += '","played":"';
						json += `${sanitize(row[10])}`;
						// map results
						json += `","map1":{"name":"${sanitize(row[11])}","winner":"${sanitize(row[12])}"},`;
						json += `"map2":{"name":"${sanitize(row[13])}","winner":"${sanitize(row[14])}"},`;
						json += `"map3":{"name":"${sanitize(row[15])}","winner":"${sanitize(row[16])}"},`;
						json += `"map4":{"name":"${sanitize(row[17])}","winner":"${sanitize(row[18])}"},`;
						json += `"map5":{"name":"${sanitize(row[19])}","winner":"${sanitize(row[20])}"},`;
						json += `"map6":{"name":"${sanitize(row[21])}","winner":"${sanitize(row[22])}"},`;
						json += `"map7":{"name":"${sanitize(row[23])}","winner":"${sanitize(row[24])}"},`;
						json += `"map8":{"name":"${sanitize(row[25])}","winner":"${sanitize(row[26])}"},`;
						json += `"winner":"${sanitize(row[27])}",`;
						json += `"vod":"${sanitize(row[29])}"`;
						json += '},';
					});
					
					json = json.replace(/,$/,'');
					json += ']}';
				} else {
					json += '},';
				}
				console.log(json);
				
				// save json
				let jsonObj = JSON.parse(json);
				localfs.writeJsonPromise('./data/' + this.internal + '/tournaments/' + this.tournaments[i].internal + '.json', jsonObj)
				.then(() => {})
				.catch(err => {
					console.log("Error storing matchlog!");
					reject(err);
				});
			}
			
			resolve("Done!");
		});
	}
	
	// Following functions just grab files from the right part of the 
	// data structure defined earlier

	// promise gets the players.json file and returns it
	getPlayers() {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise('./data/' + this.internal + '/players.json')
			.then(res => resolve(res))
			.catch(err => reject(err));
		});
	}
	
	// same as getPlayers but returns a specific player.
	getPlayerInfo(battletag) {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise('./data/' + this.internal + '/players.json')
			.then(res => utils.getPlayerFromObj(res, battletag))
			.then(res2 => resolve(res2))
			.catch(err => reject(err));
		});
	}
	
	// promise gets the standings file and returns it 
	getStandings() {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise('./data/' + this.internal + '/standings.json')
			.then(res => resolve(res))
			.catch(err => reject(err));
		});
	}
	
	// promise gets team's info and returns it
	getTeamInfo(team) {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise("./data/" + this.internal + "/teams/" + team + ".json")
			.then(data => resolve(data))
			.catch(err => reject(err));
		});
	}
	
	// promise gets the match log and returns it
	getMatches() {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise("./data/" + this.internal + "/matchlog.json")
			.then(res => resolve(res))
			.catch(err => reject(err));
		});
	}
	
	// promise gets the hero stats and returns it
	getHeroStats() {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise("./data/" + this.internal + "/herostats.json")
			.then(res => resolve(res))
			.catch(err => reject(err));
		});
	}

	getTournamentInfo(tournament) {
		return new Promise((resolve,reject) => {
			localfs.openJsonPromise("./data/" + this.internal + "/tournaments/" + tournament + ".json")
			.then(data => resolve(data))
			.catch(err => reject(err));
		});
	}
}
